File: aeronav_download.py
#!/usr/bin/env python3
'''
Standalone downloader for FAA Aeronav chart data.

Downloads current GeoTIFF chart ZIP files from aeronav.faa.gov.
Supports incremental updates via If-Modified-Since headers.

Usage:
    python aeronav_download.py <zippath>
    python aeronav_download.py --zippath /path/to/zips
    python aeronav_download.py --zippath /path/to/zips --quiet
'''

import argparse
import os
import re
import time
import urllib.error
import urllib.request

import bs4

# FAA Aeronav index URLs
_AERONAV_VFR_URL = 'https://www.faa.gov/air_traffic/flight_info/aeronav/digital_products/vfr/'
_AERONAV_IFR_URL = 'https://www.faa.gov/air_traffic/flight_info/aeronav/digital_products/ifr/'

# Chart type IDs used on the FAA website (div IDs in the HTML)
_VFR_CHART_TYPES = ['sectional', 'terminalArea', 'helicopter', 'grandCanyon', 'Planning', 'caribbean']
_IFR_CHART_TYPES = ['lowsHighsAreas', 'planning', 'caribbean', 'gulf']


def download(url, path='.', filename=None):
    '''
    Downloads a file from the given URL and saves it to the specified path and filename.
    If the filename is not provided, the file will be saved with the basename of the URL.
    If the file already exists, the function will add an 'If-Modified-Since' header to the request
    to avoid downloading the file again if it has not been modified.

    Args:
        url (str): The URL of the file to download.
        path (str, optional): The directory path to save the file. Defaults to '.'.
        filename (str, optional): The name of the file to save. Defaults to None (uses basename of URL).

    Returns:
        str: The full path of the downloaded file.

    Raises:
        urllib.error.HTTPError: If an HTTP error occurs other than a 304 Not Modified response.
    '''
    # Create a request to retrieve the file
    request = urllib.request.Request(url)

    # Set filename
    filename = os.path.join(path, filename or os.path.basename(url))

    # Check if the file already exists on the filesystem and add the If-Modified-Since header to the request
    if os.path.exists(filename):
        last_modified_time = os.path.getmtime(filename)
        last_modified_time_str = time.strftime('%a, %d %b %Y %H:%M:%S GMT', time.gmtime(last_modified_time))
        request.add_header('If-Modified-Since', last_modified_time_str)

    # Download the file
    try:
        with urllib.request.urlopen(request) as response:
            if response.status == 200:
                with open(filename, 'wb') as f:
                    f.write(response.read())

    except urllib.error.HTTPError as e:
        # Check if the server returned a 304 Not Modified response, if so, just skip the download
        if e.code != 304:
            raise

    return filename


def get_current_aeronav_urls(index_url, chart_types):
    '''
    Scrapes the aeronav.faa.gov website for the current URLs of the specified chart types.

    Args:
        index_url (str): The URL of the index page to scrape.
        chart_types (list): List of chart type div IDs to look for on the page.

    Returns:
        list: List of GeoTIFF ZIP file URLs found on the page.
    '''
    # Read the page for scraping
    with urllib.request.urlopen(index_url) as response:
        html = response.read().decode('utf-8')

    soup = bs4.BeautifulSoup(html, 'html.parser')

    urls = []
    for chart_type in chart_types:
        # Each chart type has its own div
        chart = soup.find('div', id=chart_type)
        if chart:
            # Any table describes a set of charts
            tables = chart.find_all('table')
            for table in tables:
                # Each row in the table describes a chart
                rows = table.find_all('tr')
                for row in rows:
                    # The second cell in each row contains the current chart info
                    cells = row.find_all('td')
                    if len(cells) >= 2:
                        # Find all <a> elements in the cell
                        a_tags = cells[1].find_all('a')
                        for a_tag in a_tags:
                            # Check if the link text is 'Geo-TIFF'
                            if a_tag and a_tag.get_text(strip=True).lower() == 'geo-tiff':
                                # Finally, there it is
                                geo_tiff_url = a_tag['href']
                                urls.append(geo_tiff_url)
    return urls


def fix_faa_incorrect_urls(urls):
    '''
    Fixes incorrect URLs on the aeronav.faa.gov website by finding the most common "before date" part of the URL
    and using that as the base URL for all URLs. This is necessary because the FAA's web site doesn't always
    link to the correct file. FAA, get your shit together.

    Args:
        urls (list): List of URLs to fix.

    Returns:
        list: List of corrected URLs with consistent date paths.
    '''
    url_tuples = []
    date_pattern = re.compile(r'/(\d{2}-\d{2}-\d{4})/')

    # Examine each URL, and find the date in the URL
    for url in urls:
        match = date_pattern.search(url)
        if match:
            # Store everything up to and including the date, then everything after the date
            before_date = url[:match.end(1)]
            after_date = url[match.end(1):]
            url_tuples.append((before_date, after_date))
        else:
            raise ValueError(f"Could not find date in aeronav URL {url}")

    # All "before date" parts should be the same. Tally up the counts of each "before date" part
    baseurl_counts = {}
    for before_date, _ in url_tuples:
        if before_date not in baseurl_counts:
            baseurl_counts[before_date] = 0
        baseurl_counts[before_date] += 1

    # If they are not all the same, use the most common one
    if len(baseurl_counts) > 1:
        most_common_baseurl = max(baseurl_counts, key=lambda x: baseurl_counts[x])
        cleaned_urls = []
        for before_date, after_date in url_tuples:
            if before_date != most_common_baseurl:
                cleaned_urls.append(most_common_baseurl + after_date)
            else:
                cleaned_urls.append(before_date + after_date)
        return cleaned_urls
    else:
        return urls


def main():
    '''
    Main entry point for the standalone downloader.
    '''
    parser = argparse.ArgumentParser(
        description='Download FAA Aeronav chart data from aeronav.faa.gov.',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''
Examples:
    %(prog)s /path/to/zips
    %(prog)s --zippath /path/to/zips
    %(prog)s --zippath /path/to/zips --quiet
'''
    )
    parser.add_argument('zippath', nargs='?', help='Directory to store downloaded ZIP files.')
    parser.add_argument('--zippath', dest='zippath_opt', help='Directory to store downloaded ZIP files (alternative to positional argument).')
    parser.add_argument('--quiet', action='store_true', help='Suppress output.')
    args = parser.parse_args()

    # Allow either positional or --zippath argument
    zippath = args.zippath or args.zippath_opt
    if not zippath:
        parser.error('zippath is required (either as positional argument or --zippath)')

    quiet = args.quiet

    if not quiet:
        print('Scraping aeronav.faa.gov...')

    # Scrape all chart file URLs
    vfr_urls = get_current_aeronav_urls(_AERONAV_VFR_URL, _VFR_CHART_TYPES)
    ifr_urls = get_current_aeronav_urls(_AERONAV_IFR_URL, _IFR_CHART_TYPES)

    # aeronav.faa.gov may have some incorrect URLs, so we need to clean them up
    vfr_urls = fix_faa_incorrect_urls(vfr_urls)
    ifr_urls = fix_faa_incorrect_urls(ifr_urls)

    all_urls = vfr_urls + ifr_urls

    if not quiet:
        print(f'Found {len(all_urls)} chart files to download.')

    # Create the directory if it does not exist
    os.makedirs(zippath, exist_ok=True)

    # Download all the files
    downloaded = 0
    skipped = 0
    for url in all_urls:
        filename = os.path.basename(url)
        filepath = os.path.join(zippath, filename)
        existed = os.path.exists(filepath)
        old_mtime = os.path.getmtime(filepath) if existed else 0

        if not quiet:
            print(f'Downloading {filename}...', end=' ', flush=True)

        download(url, zippath)

        # Check if file was actually downloaded or skipped
        new_mtime = os.path.getmtime(filepath)
        if existed and new_mtime == old_mtime:
            skipped += 1
            if not quiet:
                print('(not modified)')
        else:
            downloaded += 1
            if not quiet:
                print('done')

    if not quiet:
        print(f'\nDownload complete: {downloaded} downloaded, {skipped} already up to date.')


if __name__ == '__main__':
    main()

File: src/aeronav_download.c
/*
 * aeronav_download.c - Download FAA Aeronav chart data
 *
 * Downloads current GeoTIFF chart ZIP files from aeronav.faa.gov.
 * Supports incremental updates via If-Modified-Since headers.
 *
 * Dependencies: libcurl, libxml2
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <unistd.h>
#include <sys/stat.h>
#include <time.h>
#include <getopt.h>
#include <errno.h>
#include <ctype.h>

#include <curl/curl.h>
#include <libxml/HTMLparser.h>
#include <libxml/xpath.h>

/* FAA Aeronav index URLs */
#define AERONAV_VFR_URL "https://www.faa.gov/air_traffic/flight_info/aeronav/digital_products/vfr/"
#define AERONAV_IFR_URL "https://www.faa.gov/air_traffic/flight_info/aeronav/digital_products/ifr/"

/* Maximum number of URLs we can handle */
#define MAX_URLS 256

/* Chart type div IDs on the FAA website */
static const char *VFR_CHART_TYPES[] = {
    "sectional", "terminalArea", "helicopter", "grandCanyon", "Planning", "caribbean", NULL
};
static const char *IFR_CHART_TYPES[] = {
    "lowsHighsAreas", "planning", "caribbean", "gulf", NULL
};

/* Global quiet flag */
static bool quiet = false;

/* ============================================================================
 * Curl callback
 * ============================================================================ */

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    return fwrite(contents, size, nmemb, (FILE *)userp);
}

/* ============================================================================
 * HTTP functions
 * ============================================================================ */

/*
 * Fetch a URL and return the response body.
 * Returns NULL on error. Caller must free the returned string.
 */
static char *fetch_url(CURL *curl, const char *url) {
    char *buf = NULL;
    size_t size = 0;
    FILE *stream = open_memstream(&buf, &size);
    if (!stream) {
        fprintf(stderr, "Error creating memory stream\n");
        return NULL;
    }

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, stream);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "aeronav_download/1.0");

    CURLcode res = curl_easy_perform(curl);
    fclose(stream);

    if (res != CURLE_OK) {
        fprintf(stderr, "Error fetching %s: %s\n", url, curl_easy_strerror(res));
        free(buf);
        return NULL;
    }

    return buf;
}

/*
 * Download a file with If-Modified-Since support.
 * Returns: 1 = downloaded, 0 = not modified, -1 = error
 */
static int download_file(CURL *curl, const char *url, const char *filepath) {
    struct stat st;
    bool file_exists = (stat(filepath, &st) == 0);
    struct curl_slist *headers = NULL;

    /* Set If-Modified-Since header if file exists */
    if (file_exists) {
        char timebuf[64];
        struct tm *tm = gmtime(&st.st_mtime);
        strftime(timebuf, sizeof(timebuf), "If-Modified-Since: %a, %d %b %Y %H:%M:%S GMT", tm);
        headers = curl_slist_append(headers, timebuf);
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    }

    /* Create temp file for download */
    char tmppath[4096];
    snprintf(tmppath, sizeof(tmppath), "%s.tmp", filepath);
    FILE *fp = fopen(tmppath, "wb");
    if (!fp) {
        fprintf(stderr, "Error creating %s: %s\n", tmppath, strerror(errno));
        if (headers) curl_slist_free_all(headers);
        return -1;
    }

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "aeronav_download/1.0");

    CURLcode res = curl_easy_perform(curl);
    fclose(fp);

    if (headers) {
        curl_slist_free_all(headers);
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, NULL);
    }

    if (res != CURLE_OK) {
        fprintf(stderr, "Error downloading %s: %s\n", url, curl_easy_strerror(res));
        unlink(tmppath);
        return -1;
    }

    long response_code;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);

    if (response_code == 304) {
        /* Not modified */
        unlink(tmppath);
        return 0;
    } else if (response_code == 200) {
        /* Downloaded successfully - move temp file to final location */
        if (rename(tmppath, filepath) != 0) {
            fprintf(stderr, "Error moving %s to %s: %s\n", tmppath, filepath, strerror(errno));
            unlink(tmppath);
            return -1;
        }
        return 1;
    } else {
        fprintf(stderr, "Unexpected response code %ld for %s\n", response_code, url);
        unlink(tmppath);
        return -1;
    }
}

/* ============================================================================
 * HTML parsing
 * ============================================================================ */

/*
 * Check if a string matches case-insensitively, trimming whitespace.
 */
static bool text_matches(const char *text, const char *target) {
    if (!text) return false;

    /* Skip leading whitespace */
    while (*text && isspace((unsigned char)*text)) text++;

    /* Compare case-insensitively */
    size_t target_len = strlen(target);
    if (strncasecmp(text, target, target_len) != 0) return false;

    /* Check trailing is only whitespace */
    text += target_len;
    while (*text) {
        if (!isspace((unsigned char)*text)) return false;
        text++;
    }
    return true;
}

/*
 * Get text content of a node (concatenated text of all descendants).
 */
static char *get_node_text(xmlNodePtr node) {
    xmlChar *content = xmlNodeGetContent(node);
    if (!content) return NULL;
    char *result = strdup((char *)content);
    xmlFree(content);
    return result;
}

/*
 * Recursively find anchor tags with "Geo-TIFF" text and add their hrefs to the array.
 */
static void find_geotiff_links(xmlNodePtr node, char *urls[], int *count, int max) {
    for (xmlNodePtr child = node->children; child; child = child->next) {
        if (child->type == XML_ELEMENT_NODE) {
            if (xmlStrcmp(child->name, (xmlChar *)"a") == 0) {
                /* Check if link text is "Geo-TIFF" (case-insensitive) */
                char *text = get_node_text(child);
                if (text && text_matches(text, "Geo-TIFF")) {
                    xmlChar *href = xmlGetProp(child, (xmlChar *)"href");
                    if (href) {
                        if (*count < max) {
                            urls[(*count)++] = strdup((char *)href);
                        }
                        xmlFree(href);
                    }
                }
                free(text);
            } else {
                /* Recurse into other elements */
                find_geotiff_links(child, urls, count, max);
            }
        }
    }
}

/*
 * Scrape chart URLs from an FAA index page.
 */
static void scrape_chart_urls(const char *html, const char **chart_types,
                              char *urls[], int *count, int max) {
    /* Parse HTML */
    htmlDocPtr doc = htmlReadMemory(html, strlen(html), NULL, NULL,
                                     HTML_PARSE_NOERROR | HTML_PARSE_NOWARNING);
    if (!doc) {
        fprintf(stderr, "Error parsing HTML\n");
        return;
    }

    xmlXPathContextPtr ctx = xmlXPathNewContext(doc);
    if (!ctx) {
        xmlFreeDoc(doc);
        return;
    }

    /* For each chart type div */
    for (int i = 0; chart_types[i]; i++) {
        /* Find the div with this ID */
        char xpath[256];
        snprintf(xpath, sizeof(xpath), "//div[@id='%s']//table//tr", chart_types[i]);

        xmlXPathObjectPtr result = xmlXPathEvalExpression((xmlChar *)xpath, ctx);
        if (!result) continue;

        xmlNodeSetPtr nodes = result->nodesetval;
        if (!nodes) {
            xmlXPathFreeObject(result);
            continue;
        }

        /* For each table row */
        for (int j = 0; j < nodes->nodeNr; j++) {
            xmlNodePtr row = nodes->nodeTab[j];

            /* Find all td elements in this row */
            int td_count = 0;
            xmlNodePtr second_td = NULL;
            for (xmlNodePtr child = row->children; child; child = child->next) {
                if (child->type == XML_ELEMENT_NODE &&
                    xmlStrcmp(child->name, (xmlChar *)"td") == 0) {
                    td_count++;
                    if (td_count == 2) {
                        second_td = child;
                        break;
                    }
                }
            }

            if (!second_td) continue;

            /* Recursively find all Geo-TIFF links in the second td */
            find_geotiff_links(second_td, urls, count, max);
        }

        xmlXPathFreeObject(result);
    }

    xmlXPathFreeContext(ctx);
    xmlFreeDoc(doc);
}

/* ============================================================================
 * URL fixing
 * ============================================================================ */

/*
 * Find the date pattern in a URL (e.g., "/01-23-2025/").
 * Returns pointer to start of date, or NULL if not found.
 * Sets *end to point past the date.
 */
static const char *find_date_in_url(const char *url, const char **end) {
    /* Look for pattern: /DD-DD-DDDD/ */
    const char *p = url;
    while ((p = strchr(p, '/')) != NULL) {
        p++; /* Skip the slash */
        /* Check for DD-DD-DDDD pattern */
        if (strlen(p) >= 10 &&
            isdigit((unsigned char)p[0]) && isdigit((unsigned char)p[1]) &&
            p[2] == '-' &&
            isdigit((unsigned char)p[3]) && isdigit((unsigned char)p[4]) &&
            p[5] == '-' &&
            isdigit((unsigned char)p[6]) && isdigit((unsigned char)p[7]) &&
            isdigit((unsigned char)p[8]) && isdigit((unsigned char)p[9])) {
            *end = p + 10;
            return p;
        }
    }
    return NULL;
}

/*
 * Fix incorrect URLs by normalizing dates within a range.
 * The FAA website sometimes has inconsistent dates in URLs.
 */
static void fix_faa_incorrect_urls(char *urls[], int start, int end) {
    int n = end - start;
    if (n == 0) return;

    /* Extract (before_date, after_date) for each URL */
    char **before_date = malloc(n * sizeof(char *));
    char **after_date = malloc(n * sizeof(char *));

    for (int i = 0; i < n; i++) {
        const char *date_end;
        const char *date_start = find_date_in_url(urls[start + i], &date_end);
        if (!date_start) {
            fprintf(stderr, "Warning: Could not find date in URL: %s\n", urls[start + i]);
            before_date[i] = strdup(urls[start + i]);
            after_date[i] = strdup("");
        } else {
            /* before_date includes up to and including the date */
            size_t before_len = date_end - urls[start + i];
            before_date[i] = malloc(before_len + 1);
            memcpy(before_date[i], urls[start + i], before_len);
            before_date[i][before_len] = '\0';

            after_date[i] = strdup(date_end);
        }
    }

    /* Count occurrences of each before_date */
    int max_count = 0;
    const char *most_common = NULL;

    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = 0; j < n; j++) {
            if (strcmp(before_date[i], before_date[j]) == 0) {
                count++;
            }
        }
        if (count > max_count) {
            max_count = count;
            most_common = before_date[i];
        }
    }

    /* If there are multiple different before_dates, fix them */
    if (most_common) {
        for (int i = 0; i < n; i++) {
            if (strcmp(before_date[i], most_common) != 0) {
                /* Rebuild URL with the most common base */
                free(urls[start + i]);
                size_t new_len = strlen(most_common) + strlen(after_date[i]) + 1;
                urls[start + i] = malloc(new_len);
                snprintf(urls[start + i], new_len, "%s%s", most_common, after_date[i]);
            }
        }
    }

    /* Cleanup */
    for (int i = 0; i < n; i++) {
        free(before_date[i]);
        free(after_date[i]);
    }
    free(before_date);
    free(after_date);
}

/* ============================================================================
 * Main
 * ============================================================================ */

static void usage(const char *prog) {
    fprintf(stderr, "Usage: %s [OPTIONS] <zippath>\n", prog);
    fprintf(stderr, "\n");
    fprintf(stderr, "Download FAA Aeronav chart data from aeronav.faa.gov.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "Arguments:\n");
    fprintf(stderr, "  zippath             Directory to store downloaded ZIP files\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -q, --quiet         Suppress progress output\n");
    fprintf(stderr, "  -h, --help          Show this help message\n");
}

int main(int argc, char *argv[]) {
    static struct option long_options[] = {
        {"quiet", no_argument, 0, 'q'},
        {"help",  no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "qh", long_options, NULL)) != -1) {
        switch (opt) {
            case 'q':
                quiet = true;
                break;
            case 'h':
                usage(argv[0]);
                return 0;
            default:
                usage(argv[0]);
                return 1;
        }
    }

    if (optind >= argc) {
        fprintf(stderr, "Error: zippath is required\n\n");
        usage(argv[0]);
        return 1;
    }

    const char *zippath = argv[optind];

    /* Initialize libxml2 */
    xmlInitParser();
    LIBXML_TEST_VERSION

    /* Initialize curl */
    curl_global_init(CURL_GLOBAL_DEFAULT);
    CURL *curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "Error initializing curl\n");
        return 1;
    }

    if (!quiet) {
        printf("Scraping aeronav.faa.gov...\n");
    }

    /* Single array for all URLs */
    char *urls[MAX_URLS];
    int count = 0;

    /* Fetch and parse VFR page */
    char *vfr_html = fetch_url(curl, AERONAV_VFR_URL);
    if (vfr_html) {
        scrape_chart_urls(vfr_html, VFR_CHART_TYPES, urls, &count, MAX_URLS);
        free(vfr_html);
    }
    int vfr_count = count;

    /* Fix VFR URLs */
    fix_faa_incorrect_urls(urls, 0, vfr_count);

    /* Fetch and parse IFR page */
    char *ifr_html = fetch_url(curl, AERONAV_IFR_URL);
    if (ifr_html) {
        scrape_chart_urls(ifr_html, IFR_CHART_TYPES, urls, &count, MAX_URLS);
        free(ifr_html);
    }

    /* Fix IFR URLs (only the IFR portion) */
    fix_faa_incorrect_urls(urls, vfr_count, count);

    if (!quiet) {
        printf("Found %d chart files to download.\n", count);
    }

    /* Create output directory */
    if (mkdir(zippath, 0755) != 0 && errno != EEXIST) {
        fprintf(stderr, "Error creating directory %s: %s\n", zippath, strerror(errno));
        for (int i = 0; i < count; i++) free(urls[i]);
        curl_easy_cleanup(curl);
        curl_global_cleanup();
        xmlCleanupParser();
        return 1;
    }

    /* Download all files */
    int downloaded = 0;
    int skipped = 0;
    int errors = 0;

    for (int i = 0; i < count; i++) {
        const char *url = urls[i];

        /* Extract filename from URL */
        const char *filename = strrchr(url, '/');
        filename = filename ? filename + 1 : url;

        char filepath[4096];
        snprintf(filepath, sizeof(filepath), "%s/%s", zippath, filename);

        if (!quiet) {
            printf("Downloading %s... ", filename);
            fflush(stdout);
        }

        int result = download_file(curl, url, filepath);
        if (result == 1) {
            downloaded++;
            if (!quiet) printf("done\n");
        } else if (result == 0) {
            skipped++;
            if (!quiet) printf("(not modified)\n");
        } else {
            errors++;
            /* Error message already printed */
        }
    }

    if (!quiet) {
        printf("\nDownload complete: %d downloaded, %d already up to date",
               downloaded, skipped);
        if (errors > 0) {
            printf(", %d errors", errors);
        }
        printf(".\n");
    }

    /* Cleanup */
    for (int i = 0; i < count; i++) free(urls[i]);
    curl_easy_cleanup(curl);
    curl_global_cleanup();
    xmlCleanupParser();

    return errors > 0 ? 1 : 0;
}
